### PR TITLE
First iteration of entropy metric, test functions, backend updates

### PR DESCRIPTION
**Pull request by taylorpatti**
Basic implementation (matrix and CP only) of vonNeumann entropy and corresponding test functions. Minor update of main 3 backends to accommodate new functions.

### CODE REVIEW

**Comment by taylorpatti** [comment]
Looks like an issue with how I added log2 to some of the backends (Numpy, PyTorch, and MXNet). If you know the issue off the top of your head, please let me know. If not, I can troubleshoot it a little later.

File: tensorly/metrics/entropy.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    tensor : (matrix or CP tensor)
+        Data structure
+    offset : (float)
+        Minimum value at which to not consider eigenvalue zero to avoid singularities.
+
+    Returns
+    -------
+    float
+    """
+    if tensor.__class__.__name__ == 'CPTensor':
```
Comment (merajhashemi) [inline tensorly/metrics/entropy.py:22]:
```python3
if isinstance(tensor, CPTensor):
```

File: tensorly/metrics/entropy.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        weights = tensor[0]
+        if len(weights) == 1:
+            return 0
+        if np.allclose(weights, np.ones(len(weights))):
+            weights = T.tensor(cp_normalize(tensor)[0])
+        return T.sum(-weights*T.log2(weights))
```
Comment (merajhashemi) [inline tensorly/metrics/entropy.py:28]:
```python3
return -float(T.sum(weights*T.log2(weights)))
```
Some backends return 0-dim tensors

File: tensorly/metrics/entropy.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    tensor = T.tensor(tensor)
+    eig_vals = T.eigh(tensor)[0]
+    eig_vals = eig_vals[eig_vals > offset]
+    return -T.sum(T.log2(eig_vals)*eig_vals)
```
Comment (merajhashemi) [inline tensorly/metrics/entropy.py:33]:
```python3
return - float(T.sum(T.log2(eig_vals)*eig_vals))
```

File: tensorly/metrics/tests/test_entropy.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+       [0.02577351, 0.10426328, 0.06440067, 0.02632028, 0.05930842,
+        0.05479861, 0.04390419, 0.10996953],
+       [0.09599539, 0.12271515, 0.09704334, 0.14030151, 0.03368433,
+        0.04390419, 0.15988065, 0.12634305],
+       [0.07482428, 0.2212482 , 0.14317333, 0.08937212, 0.11477836,
+        0.10996953, 0.12634305, 0.23241005]])
+
+def test_vonNeumann_entropy_pure_state():
+    tl_vne = np.array(vonNeumann_entropy(mat_pure))
+    assert np.allclose(tl_vne, np.array([0]), rtol=1e-03, atol=1e-03)
```
Comment (merajhashemi) [inline tensorly/metrics/tests/test_entropy.py:44]: We can have random tests like in your previous commit.
```python3
from ...random import random_tensor


def test_vonNeumann_entropy_pure_state():
    state = random_tensor((16, 1))
    state = state / tl.norm(state)
    mat = tl.dot(state, tl.transpose(state))
    tl_vne = vonNeumann_entropy(mat)
    np.testing.assert_almost_equal(tl_vne, 0.)

```

File: tensorly/metrics/tests/test_entropy.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+def test_vonNeumann_entropy_pure_state_CP():
+    mat = parafac(tl.tensor(mat_pure), rank=1, normalize_factors=True)
+    tl_vne = np.array(vonNeumann_entropy(mat))
+    assert np.allclose(tl_vne, np.array([0]), rtol=1e-03, atol=1e-03)
+
+def test_vonNeumann_entropy_mixed_state():
+    tl_vne = np.array(vonNeumann_entropy(mat_mixed))
+    assert np.allclose(tl_vne, np.array([0.7072100566602336]), rtol=1e-03, atol=1e-03)
```
Comment (merajhashemi) [inline tensorly/metrics/tests/test_entropy.py:53]: We can use a maximally mixed state as the test case here.
```python3
def test_vonNeumann_entropy_mixed_state():
    # maximally mixed state
    d = 16
    basis, _ = tl.qr(random_tensor((d, d)))  # get orthonormal basis
    mat = (1 / d) * tl.dot(basis, tl.transpose(basis))
    tl_vne = vonNeumann_entropy(mat)
    np.testing.assert_almost_equal(tl_vne, tl.log2(d))
```

File: tensorly/metrics/entropy.py (hunk 1 above)
Comment (merajhashemi) [reply tensorly/metrics/entropy.py:22]: Isn’t the entropy defined for a density matrix though? When will it be a higher-order tensor?

File: tensorly/backend/__init__.py [hunk 6]
```diff
@@ -224,6 +224,7 @@ def inner(*args, **kwargs):
 finfo = dispatch(Backend.finfo)
 index = Backend.index
 index_update = dispatch(Backend.index_update)
+log2 = dispatch(Backend.log2)
```
Comment (merajhashemi) [inline tensorly/backend/__init__.py:227]: We need to add `log2` to Cupy, JAX, and TensorFlow too.

File: tensorly/metrics/entropy.py (hunk 1 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:22]: We're going to sometimes organize it as a higher order tensor, at which point it will be more efficient to do the calculation in tensor form.

Comment (merajhashemi) [reply tensorly/metrics/entropy.py:22]:
```python3
    if isinstance(tensor, CPTensor):
        weights = cp_normalize(tensor).weights
        vn_entropy = -T.sum(weights * T.log2(weights))

    else:
        eig_vals = T.eigh(tensor)[0]
        eig_vals = eig_vals[eig_vals > offset]
        vn_entropy = -T.sum(T.log2(eig_vals) * eig_vals)

    return float(vn_entropy)
```

File: tensorly/metrics/tests/test_entropy.py (hunk 4 above)
Comment (taylorpatti) [reply tensorly/metrics/tests/test_entropy.py:44]: Jean requested that I hardcore an example so as to have a pre-calculated answer and not import Qutip.

File: tensorly/metrics/tests/test_entropy.py (hunk 5 above)
Comment (taylorpatti) [reply tensorly/metrics/tests/test_entropy.py:53]: Is there an advantage to changing this?

File: tensorly/backend/__init__.py (hunk 6 above)
Comment (taylorpatti) [reply tensorly/backend/__init__.py:227]: Yep, indeed. Working on that for sure.

File: tensorly/metrics/entropy.py (hunk 2 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:28]: Good to know, thanks. Under what circumstances?

File: tensorly/metrics/entropy.py (hunk 1 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:22]: How about:

    if isinstance(tensor, CPTensor):
        eig_vals = cp_normalize(tensor).weights

    else:
        eig_vals = T.eigh(tensor)[0]
        eig_vals = eig_vals[eig_vals > offset]

    vn_entropy = -T.sum(T.log2(eig_vals) * eig_vals)

    return float(vn_entropy)

File: tensorly/metrics/entropy.py (hunk 2 above)
Comment (merajhashemi) [reply tensorly/metrics/entropy.py:28]: For Pytorch, JAX, TensorFlow it will return a 0-dim tensor.

File: tensorly/metrics/tests/test_entropy.py (hunk 4 above)
Comment (merajhashemi) [reply tensorly/metrics/tests/test_entropy.py:44]: We know the entropy is zero for pure states so there's no need for `Qutip` with this example.

File: tensorly/metrics/tests/test_entropy.py (hunk 5 above)
Comment (merajhashemi) [reply tensorly/metrics/tests/test_entropy.py:53]: It just looks nicer than hardcoding the tensors!

File: tensorly/metrics/tests/test_entropy.py (hunk 4 above)
Comment (taylorpatti) [reply tensorly/metrics/tests/test_entropy.py:44]: Right, which is why before it was showing it both for QuTip and with zero itself - as an illustration. Jean requested I remove QuTip on all accounts though, which is why I hardcoded both.

File: tensorly/metrics/entropy.py (hunk 1 above)
Comment (merajhashemi) [reply tensorly/metrics/entropy.py:22]:
```python3
eig_vals = eig_vals[eig_vals > offset]
```
How about we keep this line outside the else statement?

Comment (taylorpatti) [reply tensorly/metrics/entropy.py:22]: Sounds good. I'm also consolidating the last two lines.

File: tensorly/metrics/entropy.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    tensor : (matrix or CP tensor)
+        Data structure
+    offset : (float)
+        Minimum value at which to not consider eigenvalue zero to avoid singularities.
+
+    Returns
+    -------
+    float
+    """
+    if isinstance(tensor, CPTensor):
```
Comment (JeanKossaifi) [inline tensorly/metrics/entropy.py:23]: I'm wondering whether it would be worth having an explicit cp_von_neumann_entropy?

File: tensorly/metrics/entropy.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+from .. import backend as T
+from ..cp_tensor import CPTensor
+from ..cp_tensor import cp_normalize
+import numpy as np
+
+# Author: Taylor Lee Patti <taylorpatti@g.harvard.edu>
+
+
+def vonNeumann_entropy(tensor, offset=1e-12):
```
Comment (JeanKossaifi) [inline tensorly/metrics/entropy.py:9]: I would tend to use von_neumann_entropy for a function and VonNeumannEntropy for a class.

File: tensorly/metrics/entropy.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    -------
+    float
+    """
+    if isinstance(tensor, CPTensor):
+        eig_vals = cp_normalize(tensor).weights
+
+    else:
+        eig_vals = T.eigh(tensor)[0]
+
+    eig_vals = eig_vals[eig_vals > offset]
```
Comment (JeanKossaifi) [inline tensorly/metrics/entropy.py:29]: This offset will only be valid for float64, we can use the dtype's machine precision instead: `eps = tl.eps(tensor.dtype)`.

File: tensorly/metrics/entropy.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+
+    return float(-T.sum(T.log2(eig_vals) * eig_vals))
```
Comment (JeanKossaifi) [inline tensorly/metrics/entropy.py:32]: Not sure we really want to cast to float, this will break the computational graph if we're e.g. tracking gradients.

File: tensorly/metrics/tests/test_entropy.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+import numpy.random as npr
+import numpy.linalg as npl
+import tensorly as tl
+from ..entropy import vonNeumann_entropy
+from ...decomposition import parafac
+
+state = npr.rand(8, 1).astype(np.float32)
```
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_entropy.py:8]: I would define these inside the unit-test functions. If we no longer cast to float it would also be better to use tl.randn and tl.tensor so the arrays are in created with the right backend.

File: tensorly/metrics/tests/test_entropy.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+state = state/npl.norm(state)
+mat_pure = tl.tensor(np.matmul(state, np.transpose(state)))
```
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_entropy.py:10]:
```suggestion
mat_pure = tl.dot(state, tl.transpose(state))
```

File: tensorly/metrics/tests/test_entropy.py (hunk 11 above)
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_entropy.py:8]:
```suggestion
state = tl.tensor(npr.rand(8, 1), dtype=tl.float32)
```

File: tensorly/metrics/tests/test_entropy.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
+import numpy as np
+import numpy.random as npr
+import numpy.linalg as npl
+import tensorly as tl
+from ..entropy import vonNeumann_entropy
+from ...decomposition import parafac
+
+state = npr.rand(8, 1).astype(np.float32)
+state = state/npl.norm(state)
```
Comment (JeanKossaifi) [inline tensorly/metrics/tests/test_entropy.py:9]:
```suggestion
state = state/tl.norm(state)
```

File: tensorly/metrics/tests/test_entropy.py (hunk 4 above)
Comment (JeanKossaifi) [reply tensorly/metrics/tests/test_entropy.py:20]: We can also have both: a random test + deterministic one.
If all are tl.tensors, we can test using `tensorly.testing`, e.g. `tensorly.testing.assert_array_almost_equal`.

**Review by JeanKossaifi — CHANGES_REQUESTED** [review]
Thanks @taylorpatti for the PR and @merajhashemi for the comments!

Looks good, I'd just make it fully backend agnostic - in general it's best to not convert to numpy whenever possible as this is the most flexible for end users.

File: tensorly/metrics/entropy.py (hunk 10 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:32]: Yeah, I disagreed with this as well.

File: tensorly/metrics/entropy.py (hunk 7 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:23]: So you would advise against autodetect for users? Or do you mean just define the main function in terms of the general, cp, and soon tt version? The later would lead to more repetitious code but

Comment (JeanKossaifi) [reply tensorly/metrics/entropy.py:23]: Right -- the issue is that now the API for the class is not fixed and mostly experimental (something we need to work on). So far I have a function cp_do_something for most operations that users can call explicitly, though I'm not sure it's the best way either. 

In general now, all cp_tensors are in the form (weights, factors) so the cp_von_neumann_entropy could be:
```
def cp_neumann_entropy(cp_tensor):
    weights, factors = cp_normalize(cp_tensor)
    ...
```
which also will allow tuples, not just CPTensors

As a minor side note, would it make sense to name it just `neumann_entropy`?

Comment (taylorpatti) [reply tensorly/metrics/entropy.py:23]: Okay, I get you, for consistency. I'll separate them then.

How about cp_vonneumann_entropy? I think the "von" makes the function's purpose clear - to my background at least.

Comment (JeanKossaifi) [reply tensorly/metrics/entropy.py:23]: I agree, let's go with `vonneumann_entropy`

File: tensorly/metrics/entropy.py (hunk 9 above)
Comment (taylorpatti) [reply tensorly/metrics/entropy.py:29]: It seems that CPTensor might not have a dtype attribute? Should I add one?

Comment (merajhashemi) [reply tensorly/metrics/entropy.py:29]: The native backend tensors have a dtype attribute. we can do something like:
```python3
eps = T.eps(eig_vals.dtpe)
```

File: tensorly/metrics/entropy.py (hunk 10 above)
Comment (merajhashemi) [reply tensorly/metrics/entropy.py:32]: I suggested this to be consistent with the docstring. We can say the function returns a scalar tensor instead of float:?

File: tensorly/metrics/entropy.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    """Returns the von Neumann entropy of the tensor.
+
+    Parameters
+    ----------
+    tensor : (matrix)
+        Data structure
+
+    Returns
+    -------
+    float
```
Comment (JeanKossaifi) [inline tensorly/metrics/entropy.py:19]:
```suggestion
    von_neumann_entropy : order-0 tensor
```

**Comment by JeanKossaifi** [comment]
Awesome, thanks a lot @taylorpatti ! Merging.
